crates/rpc/src/v03/method/get_events.rs

```diff
@@ -0,0 +1,893 @@
+            request.chunk_size,
+            is_last_page,
+        );
+    }
+
+    check_continuation_token_validity(requested_offset, &events.events)?;
+
+    Ok(events)
+}
+
+// Maps `to_block` BlockId to a block number which can be used by the events query.
+//
+// This block id specifies the upper end of the range, so pending/latest/None means
+// there's no upper limit.
+fn map_to_block_to_number(
+    tx: &rusqlite::Transaction<'_>,
+    block: Option<BlockId>,
+) -> Result<Option<StarknetBlockNumber>, GetEventsError> {
+    use BlockId::*;
+
+    match block {
+        Some(Hash(hash)) => {
+            let number =
+                StarknetBlocksTable::get_number(tx, hash)?.ok_or(GetEventsError::BlockNotFound)?;
+
+            Ok(Some(number))
+        }
+        Some(Number(number)) => Ok(Some(number)),
+        Some(Pending) | Some(Latest) | None => Ok(None),
+    }
+}
+
+// Maps `from_block` BlockId to a block number which can be used by the events query.
+//
+// This block id specifies the lower end of the range, so pending/latest means
+// a lower limit here.
+fn map_from_block_to_number(
+    tx: &rusqlite::Transaction<'_>,
+    block: Option<BlockId>,
+) -> Result<Option<StarknetBlockNumber>, GetEventsError> {
+    use BlockId::*;
+
+    match block {
+        Some(Hash(hash)) => {
+            let number =
+                StarknetBlocksTable::get_number(tx, hash)?.ok_or(GetEventsError::BlockNotFound)?;
+
+            Ok(Some(number))
+        }
+        Some(Number(number)) => Ok(Some(number)),
+        Some(Pending) | Some(Latest) => {
+            let number =
+                StarknetBlocksTable::get_latest_number(tx)?.ok_or(GetEventsError::BlockNotFound)?;
+            Ok(Some(number))
+        }
+        None => Ok(None),
+    }
+}
+
+/// Append's pending events to `dst` based on the filter requirements and returns
+/// true if this was the last pending data i.e. `is_last_page`.
+async fn append_pending_events(
+    pending_data: &Option<PendingData>,
+    dst: &mut Vec<types::EmittedEvent>,
+    skip: usize,
+    amount: usize,
+    address: Option<ContractAddress>,
+    keys: Vec<std::collections::HashSet<EventKey>>,
+) -> bool {
+    let pending_block = match pending_data.as_ref() {
+        Some(data) => match data.block().await {
+            Some(block) => block,
+            None => return true,
+        },
+        None => return true,
+    };
+
+    let original_len = dst.len();
+
+    let pending_events = pending_block
+        .transaction_receipts
+        .iter()
+        .flat_map(|receipt| {
+            receipt
+                .events
+                .iter()
+                .zip(std::iter::repeat(receipt.transaction_hash))
+        })
+        .filter(|(event, _)| match address {
+            Some(address) => event.from_address == address,
+            None => true,
+        })
+        .filter(|(event, _)| {
+            // check if filter is empty
+            if keys.iter().flatten().count() == 0 {
```

Comment:
nit: extract count from being re-calculated for each event?

Reply:
Thanks, that's a good point. I've fixed this.